Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.cpp

```diff
@@ -42,29 +42,65 @@ DisplayConfigurationMonitor::~DisplayConfigurationMonitor() = default;
 
 DisplayConfigurationMonitor& DisplayConfigurationMonitor::singleton()
 {
-    assertIsMainThread();
     static NeverDestroyed<DisplayConfigurationMonitor> instance;
     return instance;
 }
 
-void DisplayConfigurationMonitor::addClient(Client& client)
+void DisplayConfigurationMonitor::addClient(Client& client, SerialFunctionDispatcher* dispatcher)
 {
-    ASSERT(!m_clients.contains(&client));
-    m_clients.append(&client);
+    Locker locker { m_lock };
+    if (dispatcher)
+        assertIsCurrent(*dispatcher);
+    else
+        assertIsMainThread();
+    ASSERT(!m_clients.contains(ClientEntry { &client }));
+    m_clients.append({ &client, dispatcher });
 }
 
 void DisplayConfigurationMonitor::removeClient(Client& client)
 {
-    ASSERT(m_clients.contains(&client));
-    m_clients.removeFirst(&client);
+    Locker locker { m_lock };
+    size_t index = m_clients.find(ClientEntry { &client });
+    ASSERT(index != notFound);
+    if (m_clients.at(index).m_dispatcher)
+        assertIsCurrent(*m_clients.at(index).m_dispatcher);
+    else
+        assertIsMainThread();
+    m_clients.removeFirst(ClientEntry { &client });
```

Comment:
`m_clients.removeIndex(index);`, since we've already looked it up. (It will `ASSERT_WITH_SECURITY_IMPLICATIONS` with an invalid index.)